<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_07) on Tue Oct 21 19:46:59 EDT 2008 -->
<TITLE>
DataSetCore
</TITLE>

<META NAME="date" CONTENT="2008-10-21">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="DataSetCore";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataTable.html" title="class in edu.cmu.cs.sb.core"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?edu/cmu/cs/sb/core/DataSetCore.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DataSetCore.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
edu.cmu.cs.sb.core</FONT>
<BR>
Class DataSetCore</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../resources/inherit.gif" ALT="extended by "><B>edu.cmu.cs.sb.core.DataSetCore</B>
</PRE>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../../../../edu/cmu/cs/sb/drem/DREM_DataSet.html" title="class in edu.cmu.cs.sb.drem">DREM_DataSet</A>, <A HREF="../../../../../edu/cmu/cs/sb/stem/STEM_DataSet.html" title="class in edu.cmu.cs.sb.stem">STEM_DataSet</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>DataSetCore</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
The class encapsulates a set of gene expression data
<P>

<P>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#badd0">badd0</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;True if a column of inital 0's should be added to the data file, false otherwise</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#bfullrepeat">bfullrepeat</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If repeat data comes from distinct time series (true) that is a longitudal time series,
 otherwise it is false and each column of the same time point between data sets is interchangeable</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#bmaxminval">bmaxminval</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;True if gene change threshold for filtering is based on the max-min difference
 False if the gene change threshold for filtering is based on the absolute difference</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#bspotincluded">bspotincluded</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;True if the spot column was included in the data file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#btakelog">btakelog</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If the log-ratio taken (true); else data is already in log space (false)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#data">data</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The expression data, row are genes, columns are time points in experiments</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#dmincorrelation">dmincorrelation</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Minimum average pairwise correlation a gene must have between full repeats if bfullrepeat is true</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#dsamplemins">dsamplemins</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The time points at which the expression data was sampled</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#dthresholdvalue">dthresholdvalue</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The threshold value for required change</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#genenames">genenames</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The list of gene names for the current data set</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][][][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#generepeatspottimedata">generepeatspottimedata</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Present/missing data for several data sets
First dimension is gene
Second dimension is repeat
Third dimension is spot
Fourth dimension is expression value (in log ratio form against time zero)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[][][][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#generepeatspottimepma">generepeatspottimepma</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Present/missing data for several data sets
First dimension is gene
Second dimension is repeat
Third dimension is spot
Fourth dimension is present/missing value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#genespottimedata">genespottimedata</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Present/missing data for one data set
First dimension is gene
Second dimension is spot
Third is expression value (in log ratio form against time zero)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[][][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#genespottimepma">genespottimepma</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Present/missing data for one data set
First dimension is gene
Second dimension is spot
Third is present/missing value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#htFiltered">htFiltered</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Contains genes filtered.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#nmaxmissing">nmaxmissing</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The maximum number of missing values to prevent a gene from being filtered.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#numcols">numcols</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Number of columns in the data matrix.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#numrows">numrows</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Number of rows in the data matrix.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#otherInputFiles">otherInputFiles</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The names of the other repeat files</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#pmavalues">pmavalues</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0 if data value is missing
non-zero if present</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#probenames">probenames</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The list of probe IDs in the current data set</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#sortedcorrvals">sortedcorrvals</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The distribution of all the average pairwise correlations of genes across full repeats</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#szGeneHeader">szGeneHeader</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The header string for the gene name column</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#szInputFile">szInputFile</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The designated main data file associated with the set, others are repeats</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#szProbeHeader">szProbeHeader</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The header string for the spot ID column</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#DataSetCore()">DataSetCore</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Empty constructor</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#DataSetCore(edu.cmu.cs.sb.core.DataSetCore)">DataSetCore</A></B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>&nbsp;theDataSetCore)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor copies each field</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#addExtraToFilter(edu.cmu.cs.sb.core.GoAnnotations)">addExtraToFilter</A></B>(<A HREF="../../../../../edu/cmu/cs/sb/core/GoAnnotations.html" title="class in edu.cmu.cs.sb.core">GoAnnotations</A>&nbsp;tga)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add those genes from tga.extragenes that were filtered to thtFiltered</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#averageAndFilterDuplicates()">averageAndFilterDuplicates</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Removes duplicate gene rows in the data file and combines there values using the median</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#dataSetReader(java.lang.String, int, double, double, boolean, boolean, boolean, boolean)">dataSetReader</A></B>(java.lang.String&nbsp;szInputFile,
              int&nbsp;nmaxmissing,
              double&nbsp;dthresholdvalue,
              double&nbsp;dmincorrelation,
              boolean&nbsp;btakelog,
              boolean&nbsp;bspotincluded,
              boolean&nbsp;brepeatset,
              boolean&nbsp;badd0)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in the datafile stored in szInputFile</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filterdistprofiles(edu.cmu.cs.sb.core.DataSetCore, edu.cmu.cs.sb.core.DataSetCore[])">filterdistprofiles</A></B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>&nbsp;theDataSet1,
                   <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>[]&nbsp;RepeatSet)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Computes the average pairwise correlation between gene repeats
stores it in sortedcorrvals.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filterDuplicates()">filterDuplicates</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Removes those rows which are the duplicate of another row.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filtergenesgeneral(boolean[], int, boolean)">filtergenesgeneral</A></B>(boolean[]&nbsp;keepgene,
                   int&nbsp;nkeep,
                   boolean&nbsp;bstore)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filters those rows which do not have a true in keepgene
nkeep is the number of true rows in keepgene
If bstore is true and gene is filtered then we stroe the gene
and proble list for it in htFiltered
Returns a new DataSetCore object with those rows filtered</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filtergenesthreshold1point()">filtergenesthreshold1point</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filters those genes with expression below dthresholdvalue</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filtergenesthreshold2()">filtergenesthreshold2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If bmaxminval is true, then
filter those genes for which the difference between the max and min 
value is less than dthresholdvalue
If bmaxminval if false, then 
filter those genes for which the absolute expression change is less
than dmaxval</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filterMissing()">filterMissing</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter those rows which have nmaxmissing or more missing values or are missing the first time point value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#filterMissing1point()">filterMissing1point</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filters those rows that have a missing value at the first time point</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#logratio2()">logratio2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Converts data into log-ratio versus the first time point.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html#mergeDataSets(edu.cmu.cs.sb.core.DataSetCore[])">mergeDataSets</A></B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>[]&nbsp;otherDataSets)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given an array of otherDataSets, merges it with the current data set
by storing in data the median of the values.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="szInputFile"><!-- --></A><H3>
szInputFile</H3>
<PRE>
public java.lang.String <B>szInputFile</B></PRE>
<DL>
<DD>The designated main data file associated with the set, others are repeats
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="bfullrepeat"><!-- --></A><H3>
bfullrepeat</H3>
<PRE>
public boolean <B>bfullrepeat</B></PRE>
<DL>
<DD>If repeat data comes from distinct time series (true) that is a longitudal time series,
 otherwise it is false and each column of the same time point between data sets is interchangeable
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="otherInputFiles"><!-- --></A><H3>
otherInputFiles</H3>
<PRE>
public java.lang.String[] <B>otherInputFiles</B></PRE>
<DL>
<DD>The names of the other repeat files
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="htFiltered"><!-- --></A><H3>
htFiltered</H3>
<PRE>
public java.util.HashMap <B>htFiltered</B></PRE>
<DL>
<DD>Contains genes filtered.  Maps these gene names to the list of probe IDs associated with it
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="nmaxmissing"><!-- --></A><H3>
nmaxmissing</H3>
<PRE>
public int <B>nmaxmissing</B></PRE>
<DL>
<DD>The maximum number of missing values to prevent a gene from being filtered.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="dmincorrelation"><!-- --></A><H3>
dmincorrelation</H3>
<PRE>
public double <B>dmincorrelation</B></PRE>
<DL>
<DD>Minimum average pairwise correlation a gene must have between full repeats if bfullrepeat is true
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="numrows"><!-- --></A><H3>
numrows</H3>
<PRE>
public int <B>numrows</B></PRE>
<DL>
<DD>Number of rows in the data matrix. This corresponds to the number of genes.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="numcols"><!-- --></A><H3>
numcols</H3>
<PRE>
public int <B>numcols</B></PRE>
<DL>
<DD>Number of columns in the data matrix. This corresponds to the number of time points
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="data"><!-- --></A><H3>
data</H3>
<PRE>
public double[][] <B>data</B></PRE>
<DL>
<DD>The expression data, row are genes, columns are time points in experiments
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="pmavalues"><!-- --></A><H3>
pmavalues</H3>
<PRE>
public int[][] <B>pmavalues</B></PRE>
<DL>
<DD>0 if data value is missing
non-zero if present
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="bspotincluded"><!-- --></A><H3>
bspotincluded</H3>
<PRE>
public boolean <B>bspotincluded</B></PRE>
<DL>
<DD>True if the spot column was included in the data file
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="genespottimepma"><!-- --></A><H3>
genespottimepma</H3>
<PRE>
public int[][][] <B>genespottimepma</B></PRE>
<DL>
<DD>Present/missing data for one data set
First dimension is gene
Second dimension is spot
Third is present/missing value
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="generepeatspottimepma"><!-- --></A><H3>
generepeatspottimepma</H3>
<PRE>
public int[][][][] <B>generepeatspottimepma</B></PRE>
<DL>
<DD>Present/missing data for several data sets
First dimension is gene
Second dimension is repeat
Third dimension is spot
Fourth dimension is present/missing value
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="genespottimedata"><!-- --></A><H3>
genespottimedata</H3>
<PRE>
public double[][][] <B>genespottimedata</B></PRE>
<DL>
<DD>Present/missing data for one data set
First dimension is gene
Second dimension is spot
Third is expression value (in log ratio form against time zero)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="generepeatspottimedata"><!-- --></A><H3>
generepeatspottimedata</H3>
<PRE>
public double[][][][] <B>generepeatspottimedata</B></PRE>
<DL>
<DD>Present/missing data for several data sets
First dimension is gene
Second dimension is repeat
Third dimension is spot
Fourth dimension is expression value (in log ratio form against time zero)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="btakelog"><!-- --></A><H3>
btakelog</H3>
<PRE>
public boolean <B>btakelog</B></PRE>
<DL>
<DD>If the log-ratio taken (true); else data is already in log space (false)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="probenames"><!-- --></A><H3>
probenames</H3>
<PRE>
public java.lang.String[] <B>probenames</B></PRE>
<DL>
<DD>The list of probe IDs in the current data set
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="genenames"><!-- --></A><H3>
genenames</H3>
<PRE>
public java.lang.String[] <B>genenames</B></PRE>
<DL>
<DD>The list of gene names for the current data set
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="sortedcorrvals"><!-- --></A><H3>
sortedcorrvals</H3>
<PRE>
public double[] <B>sortedcorrvals</B></PRE>
<DL>
<DD>The distribution of all the average pairwise correlations of genes across full repeats
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="dsamplemins"><!-- --></A><H3>
dsamplemins</H3>
<PRE>
public java.lang.String[] <B>dsamplemins</B></PRE>
<DL>
<DD>The time points at which the expression data was sampled
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="dthresholdvalue"><!-- --></A><H3>
dthresholdvalue</H3>
<PRE>
public double <B>dthresholdvalue</B></PRE>
<DL>
<DD>The threshold value for required change
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="bmaxminval"><!-- --></A><H3>
bmaxminval</H3>
<PRE>
public boolean <B>bmaxminval</B></PRE>
<DL>
<DD>True if gene change threshold for filtering is based on the max-min difference
 False if the gene change threshold for filtering is based on the absolute difference
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="badd0"><!-- --></A><H3>
badd0</H3>
<PRE>
public boolean <B>badd0</B></PRE>
<DL>
<DD>True if a column of inital 0's should be added to the data file, false otherwise
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="szProbeHeader"><!-- --></A><H3>
szProbeHeader</H3>
<PRE>
public java.lang.String <B>szProbeHeader</B></PRE>
<DL>
<DD>The header string for the spot ID column
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="szGeneHeader"><!-- --></A><H3>
szGeneHeader</H3>
<PRE>
public java.lang.String <B>szGeneHeader</B></PRE>
<DL>
<DD>The header string for the gene name column
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DataSetCore()"><!-- --></A><H3>
DataSetCore</H3>
<PRE>
public <B>DataSetCore</B>()</PRE>
<DL>
<DD>Empty constructor
<P>
</DL>
<HR>

<A NAME="DataSetCore(edu.cmu.cs.sb.core.DataSetCore)"><!-- --></A><H3>
DataSetCore</H3>
<PRE>
public <B>DataSetCore</B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>&nbsp;theDataSetCore)</PRE>
<DL>
<DD>Constructor copies each field
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="addExtraToFilter(edu.cmu.cs.sb.core.GoAnnotations)"><!-- --></A><H3>
addExtraToFilter</H3>
<PRE>
public void <B>addExtraToFilter</B>(<A HREF="../../../../../edu/cmu/cs/sb/core/GoAnnotations.html" title="class in edu.cmu.cs.sb.core">GoAnnotations</A>&nbsp;tga)</PRE>
<DL>
<DD>Add those genes from tga.extragenes that were filtered to thtFiltered
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="dataSetReader(java.lang.String, int, double, double, boolean, boolean, boolean, boolean)"><!-- --></A><H3>
dataSetReader</H3>
<PRE>
protected void <B>dataSetReader</B>(java.lang.String&nbsp;szInputFile,
                             int&nbsp;nmaxmissing,
                             double&nbsp;dthresholdvalue,
                             double&nbsp;dmincorrelation,
                             boolean&nbsp;btakelog,
                             boolean&nbsp;bspotincluded,
                             boolean&nbsp;brepeatset,
                             boolean&nbsp;badd0)
                      throws java.io.IOException,
                             java.io.FileNotFoundException,
                             java.lang.IllegalArgumentException</PRE>
<DL>
<DD>Reads in the datafile stored in szInputFile
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE>java.io.FileNotFoundException</CODE>
<DD><CODE>java.lang.IllegalArgumentException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="filterMissing1point()"><!-- --></A><H3>
filterMissing1point</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filterMissing1point</B>()</PRE>
<DL>
<DD>Filters those rows that have a missing value at the first time point
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filterMissing()"><!-- --></A><H3>
filterMissing</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filterMissing</B>()</PRE>
<DL>
<DD>Filter those rows which have nmaxmissing or more missing values or are missing the first time point value
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filterDuplicates()"><!-- --></A><H3>
filterDuplicates</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filterDuplicates</B>()</PRE>
<DL>
<DD>Removes those rows which are the duplicate of another row. Stores in htgenenames for each gene
name the index of all rows associated with it
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="averageAndFilterDuplicates()"><!-- --></A><H3>
averageAndFilterDuplicates</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>averageAndFilterDuplicates</B>()</PRE>
<DL>
<DD>Removes duplicate gene rows in the data file and combines there values using the median
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="logratio2()"><!-- --></A><H3>
logratio2</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>logratio2</B>()</PRE>
<DL>
<DD>Converts data into log-ratio versus the first time point.
If btakelog is true then this the log base 2 of a value over the
time point 0 value. If it is false then this is the difference
with the time point 0 value.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="mergeDataSets(edu.cmu.cs.sb.core.DataSetCore[])"><!-- --></A><H3>
mergeDataSets</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>mergeDataSets</B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>[]&nbsp;otherDataSets)</PRE>
<DL>
<DD>Given an array of otherDataSets, merges it with the current data set
by storing in data the median of the values. If bfullrepeat is
true then stores the repeat and missing data into generepeatspottimedata and
generepeatspottimepma
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filterdistprofiles(edu.cmu.cs.sb.core.DataSetCore, edu.cmu.cs.sb.core.DataSetCore[])"><!-- --></A><H3>
filterdistprofiles</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filterdistprofiles</B>(<A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>&nbsp;theDataSet1,
                                      <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A>[]&nbsp;RepeatSet)</PRE>
<DL>
<DD>Computes the average pairwise correlation between gene repeats
stores it in sortedcorrvals. Filters those genes which do not have
a sortedcorrvals exceeding dmincorrelation
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filtergenesgeneral(boolean[], int, boolean)"><!-- --></A><H3>
filtergenesgeneral</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filtergenesgeneral</B>(boolean[]&nbsp;keepgene,
                                      int&nbsp;nkeep,
                                      boolean&nbsp;bstore)</PRE>
<DL>
<DD>Filters those rows which do not have a true in keepgene
nkeep is the number of true rows in keepgene
If bstore is true and gene is filtered then we stroe the gene
and proble list for it in htFiltered
Returns a new DataSetCore object with those rows filtered
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filtergenesthreshold2()"><!-- --></A><H3>
filtergenesthreshold2</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filtergenesthreshold2</B>()</PRE>
<DL>
<DD>If bmaxminval is true, then
filter those genes for which the difference between the max and min 
value is less than dthresholdvalue
If bmaxminval if false, then 
filter those genes for which the absolute expression change is less
than dmaxval
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filtergenesthreshold1point()"><!-- --></A><H3>
filtergenesthreshold1point</H3>
<PRE>
public <A HREF="../../../../../edu/cmu/cs/sb/core/DataSetCore.html" title="class in edu.cmu.cs.sb.core">DataSetCore</A> <B>filtergenesthreshold1point</B>()</PRE>
<DL>
<DD>Filters those genes with expression below dthresholdvalue
<P>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../edu/cmu/cs/sb/core/DataTable.html" title="class in edu.cmu.cs.sb.core"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?edu/cmu/cs/sb/core/DataSetCore.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DataSetCore.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
